swap/simulations_test.go

```diff
@@ -0,0 +1,695 @@
+		maxMsgPrice: 10000,
+		minMsgPrice: 100,
+		nodeCount:   nodeCount,
+	}
+	keys := make(map[int]*ecdsa.PrivateKey)
+	addrs := make(map[int]common.Address)
+	alloc := core.GenesisAlloc{}
+	stores := make(map[int]*state.DBStore)
+
+	// for each node, generate keys, a GenesisAccount and a state store
+	for i := 0; i < nodeCount; i++ {
+		key, err := crypto.GenerateKey()
+		if err != nil {
+			return nil, err
+		}
+		keys[i] = key
+		addrs[i] = crypto.PubkeyToAddress(key.PublicKey)
+		alloc[addrs[i]] = core.GenesisAccount{Balance: big.NewInt(10000000000)}
+		dir, err := ioutil.TempDir("", fmt.Sprintf("swap_test_store_%x", addrs[i].Hex()))
+		if err != nil {
+			return nil, err
+		}
+		stateStore, err := state.NewDBStore(dir)
+		if err != nil {
+			return nil, err
+		}
+		params.dirs[i] = dir
+		stores[i] = stateStore
+	}
+	// then create the single SimulatedBackend
+	gasLimit := uint64(8000000000)
+	defaultBackend := backends.NewSimulatedBackend(alloc, gasLimit)
+	testBackend := &swapTestBackend{SimulatedBackend: defaultBackend}
+	// finally, create all Swap instances for each node, which share the same backend
+	var owner *Owner
+	defParams := newDefaultParams()
+	for i := 0; i < nodeCount; i++ {
+		owner = createOwner(keys[i])
+		params.swaps[i] = new(stores[i], owner, testBackend, defParams)
+	}
+
+	params.backend = testBackend
+	return params, nil
+}
+
+// TestPingPongChequeSimulation just launches two nodes and sends each other
+// messages which immediately crosses the PaymentThreshold and triggers cheques
+// to each other. Checks that accounting and cheque handling works across multiple
+// cheques and also if cheques are mutually sent
+func TestPingPongChequeSimulation(t *testing.T) {
+	nodeCount := 2
+	// create the shared backend and params
+	params, err := newSharedBackendSwaps(nodeCount)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// cleanup backend
+	defer params.backend.Close()
+
+	// setup the wait for mined transaction function for testing
+	cleanup := setupContractTest()
+	defer cleanup()
+
+	params.backend.cashDone = make(chan struct{}, 1)
+	defer close(params.backend.cashDone)
+	// initialize the simulation
+	sim := simulation.NewBzzInProc(newSimServiceMap(params), false)
+	defer sim.Close()
+
+	log.Info("Initializing")
+
+	ctx, cancelSimRun := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancelSimRun()
+
+	_, err = sim.AddNodesAndConnectFull(nodeCount)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	p1 := sim.UpNodeIDs()[0]
+	p2 := sim.UpNodeIDs()[1]
+	p1Item := sim.MustNodeItem(p1, bucketKeySwap)
+	ts1 := p1Item.(*testService)
+
+	p2Item := sim.MustNodeItem(p2, bucketKeySwap)
+	ts2 := p2Item.(*testService)
+
+	for {
+		// let's always be nice and allow a time out to be catched
+		select {
+		case <-ctx.Done():
+			t.Fatal("Timed out waiting for all swap peer connections to be established")
+		default:
+		}
+		// the node has all other peers in its peer list
+		if len(ts1.swap.peers) == 1 && len(ts2.swap.peers) == 1 {
```

Comment:
Protect ts1.swap.peers and ts2.swap.peers with swap.peersLock to avoid data races.